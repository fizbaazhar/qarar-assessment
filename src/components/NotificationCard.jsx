import React from 'react';
import systemUpdateIcon from '../assets/icons/system-update.svg';
import messageIcon from '../assets/icons/message.svg';
import { NOTIFICATION_TYPES } from '../config/notificationTypes';

const NotificationCard = ({ 
  notification, 
  onToggleRead, 
  onDelete, 
  formatTimeAgo
}) => {
  const getNotificationIcon = (type) => {
    switch (type) {
      case 'comment':
        return (
          <div className="w-10 h-10 bg-[#64A4F4] rounded-lg flex items-center justify-center mr-3">
            <img src={messageIcon} alt="New Comment" className="w-5 h-5" />
          </div>
        );
      case 'system':
        return (
          <div className="w-10 h-10 bg-[#F6A723] rounded-lg flex items-center justify-center mr-3">
            <img src={systemUpdateIcon} alt="System Update" className="w-6 h-6" />
          </div>
        );
      case 'password':
      case 'update':
        return (
          <div className="w-10 h-10 bg-[#4AC29C] rounded-lg flex items-center justify-center mr-3">
            <svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path d="M16.6666 5L7.49992 14.1667L3.33325 10" stroke="white" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round"/>
            </svg>
          </div>
        );
      default:
        // Fallback to checkmark icon for any unhandled types
        return (
          <div className="w-10 h-10 bg-[#4AC29C] rounded-lg flex items-center justify-center mr-3">
            <svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path d="M16.6666 5L7.49992 14.1667L3.33325 10" stroke="white" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round"/>
            </svg>
          </div>
        );
    }
  };

  return (
    <div className="py-3 first:pt-0">
      <div className="flex items-start">
        {getNotificationIcon(notification.type)}
        <div className="flex-1">
          <h3 className="text-[15px] font-medium text-gray-900 leading-5">
            {notification.title}
          </h3>
          <p className="text-sm text-gray-500 mt-0.5">
            {formatTimeAgo(notification.timestamp)}
          </p>
        </div>
        <button
          onClick={() => onDelete(notification.id)}
          className="p-2 text-gray-400 hover:text-gray-600 rounded-full transition-colors"
          title="Delete notification"
        >
          <svg className="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M6 18L18 6M6 6l12 12" />
          </svg>
        </button>
      </div>
    </div>
  );
};

export default NotificationCard; 