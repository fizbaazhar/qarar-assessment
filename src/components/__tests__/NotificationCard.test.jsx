import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest'
import NotificationCard from '../NotificationCard'

// Mock the SVG imports
vi.mock('../assets/icons/system-update.svg', () => ({
  default: 'system-update-icon'
}))
vi.mock('../assets/icons/message.svg', () => ({
  default: 'message-icon'
}))
vi.mock('../assets/icons/tick.svg', () => ({
  default: 'tick-icon'
}))

describe('NotificationCard Component', () => {
  const mockNotification = {
    id: '1',
    title: 'Test notification',
    message: 'This is a test notification',
    type: 'info',
    timestamp: new Date().toISOString(),
    isRead: false
  }

  const mockProps = {
    notification: mockNotification,
    onToggleRead: vi.fn(),
    onDelete: vi.fn(),
    formatTimeAgo: vi.fn(() => '2 minutes ago')
  }

  // Test basic rendering
  it('renders notification with title and timestamp', () => {
    render(<NotificationCard {...mockProps} />)
    
    expect(screen.getByText('Test notification')).toBeInTheDocument()
    expect(screen.getByText('2 minutes ago')).toBeInTheDocument()
  })

  // Test different notification types
  it('renders info notification with correct icon and styling', () => {
    render(<NotificationCard {...mockProps} />)
    
    const icon = screen.getByAltText('info notification')
    expect(icon).toBeInTheDocument()
    // The icon is now an SVG data URL, so we check for the alt text instead
    expect(icon).toHaveAttribute('alt', 'info notification')
  })

  it('renders warning notification with correct icon and styling', () => {
    const warningNotification = {
      ...mockNotification,
      type: 'warning',
      title: 'Warning notification'
    }
    
    render(<NotificationCard {...mockProps} notification={warningNotification} />)
    
    const icon = screen.getByAltText('warning notification')
    expect(icon).toBeInTheDocument()
    expect(icon).toHaveAttribute('alt', 'warning notification')
  })

  it('renders success notification with correct icon and styling', () => {
    const successNotification = {
      ...mockNotification,
      type: 'success',
      title: 'Success notification'
    }
    
    render(<NotificationCard {...mockProps} notification={successNotification} />)
    
    const icon = screen.getByAltText('success notification')
    expect(icon).toBeInTheDocument()
    expect(icon).toHaveAttribute('alt', 'success notification')
  })

  // Test read/unread states
  it('shows unread indicator when notification is unread', () => {
    render(<NotificationCard {...mockProps} />)
    
    const unreadDot = screen.getByTestId('unread-indicator')
    expect(unreadDot).toHaveClass('bg-blue-500')
  })

  it('hides unread indicator when notification is read', () => {
    const readNotification = { ...mockNotification, isRead: true }
    render(<NotificationCard {...mockProps} notification={readNotification} />)
    
    const unreadDot = screen.queryByTestId('unread-indicator')
    expect(unreadDot).not.toBeInTheDocument()
  })

  it('applies different opacity for read notifications', () => {
    const readNotification = { ...mockNotification, isRead: true }
    render(<NotificationCard {...mockProps} notification={readNotification} />)
    // The outermost card div has the opacity class
    const card = screen.getByText('Test notification').closest('.py-2')
    expect(card).toHaveClass('opacity-75')
  })

  it('applies full opacity for unread notifications', () => {
    render(<NotificationCard {...mockProps} />)
    // The outermost card div has the opacity class
    const card = screen.getByText('Test notification').closest('.py-2')
    expect(card).toHaveClass('opacity-100')
  })

  // Test toggle read/unread functionality
  it('calls onToggleRead when toggle button is clicked', async () => {
    const user = userEvent.setup()
    render(<NotificationCard {...mockProps} />)
    
    const toggleButton = screen.getByTitle('Mark as read')
    await user.click(toggleButton)
    
    expect(mockProps.onToggleRead).toHaveBeenCalledWith(mockNotification)
  })

  it('shows correct toggle button icon for unread notification', () => {
    render(<NotificationCard {...mockProps} />)
    
    const toggleButton = screen.getByTitle('Mark as read')
    expect(toggleButton).toBeInTheDocument()
  })

  it('shows correct toggle button icon for read notification', () => {
    const readNotification = { ...mockNotification, isRead: true }
    render(<NotificationCard {...mockProps} notification={readNotification} />)
    
    const toggleButton = screen.getByTitle('Mark as unread')
    expect(toggleButton).toBeInTheDocument()
  })

  // Test delete functionality
  it('calls onDelete when delete button is clicked', async () => {
    const user = userEvent.setup()
    render(<NotificationCard {...mockProps} />)
    
    const deleteButton = screen.getByTitle('Delete notification')
    await user.click(deleteButton)
    
    expect(mockProps.onDelete).toHaveBeenCalledWith(mockNotification.id)
  })

  // Test accessibility
  it('provides proper accessibility attributes', () => {
    render(<NotificationCard {...mockProps} />)
    
    const toggleButton = screen.getByTitle('Mark as read')
    const deleteButton = screen.getByTitle('Delete notification')
    
    expect(toggleButton).toHaveAttribute('title', 'Mark as read')
    expect(deleteButton).toHaveAttribute('title', 'Delete notification')
  })

  // Test formatTimeAgo integration
  it('calls formatTimeAgo with notification timestamp', () => {
    const formatTimeAgo = vi.fn(() => '5 minutes ago')
    render(<NotificationCard {...mockProps} formatTimeAgo={formatTimeAgo} />)
    
    expect(formatTimeAgo).toHaveBeenCalledWith(mockNotification.timestamp)
    expect(screen.getByText('5 minutes ago')).toBeInTheDocument()
  })

  // Test error notification type
  it('renders error notification with correct styling', () => {
    const errorNotification = {
      ...mockNotification,
      type: 'error',
      title: 'Error notification'
    }
    
    render(<NotificationCard {...mockProps} notification={errorNotification} />)
    
    const icon = screen.getByAltText('error notification')
    expect(icon).toBeInTheDocument()
    expect(icon).toHaveAttribute('alt', 'error notification')
  })

  // Test unknown notification type
  it('renders unknown notification type with default styling', () => {
    const unknownNotification = {
      ...mockNotification,
      type: 'unknown',
      title: 'Unknown notification'
    }
    
    render(<NotificationCard {...mockProps} notification={unknownNotification} />)
    
    const icon = screen.getByAltText('unknown notification')
    expect(icon).toBeInTheDocument()
    expect(icon).toHaveAttribute('alt', 'unknown notification')
  })

  // Test hover states
  it('applies hover styles to toggle button', () => {
    render(<NotificationCard {...mockProps} />)
    
    const toggleButton = screen.getByTitle('Mark as read')
    expect(toggleButton).toHaveClass('hover:text-blue-600')
  })

  it('applies hover styles to delete button', () => {
    render(<NotificationCard {...mockProps} />)
    
    const deleteButton = screen.getByTitle('Delete notification')
    expect(deleteButton).toHaveClass('hover:text-red-500')
  })

  // Test snapshot
  it('matches snapshot for unread notification', () => {
    const { container } = render(<NotificationCard {...mockProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('matches snapshot for read notification', () => {
    const readNotification = { ...mockNotification, isRead: true }
    const { container } = render(
      <NotificationCard {...mockProps} notification={readNotification} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
}) 