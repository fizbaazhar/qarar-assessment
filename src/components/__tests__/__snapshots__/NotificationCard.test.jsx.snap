// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationCard Component > matches snapshot for read notification 1`] = `
<div
  class="py-2 sm:py-3 first:pt-0 transition-all duration-200 opacity-75"
>
  <div
    class="flex items-start"
  >
    <div
      class="lg:w-10 lg:h-10 md:w-8 md:h-8 w-6 h-6 rounded-lg flex items-center justify-center mr-3"
      style="background-color: rgb(100, 164, 244);"
    >
      <img
        alt="info notification"
        class="lg:w-5 lg:h-5 md:w-4 md:h-4 w-3 h-3"
        src="data:image/svg+xml,%3c?xml%20version='1.0'%20encoding='UTF-8'?%3e%3csvg%20width='20'%20height='20'%20viewBox='0%200%2020%2020'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M16.6667%201.66667H3.33333C2.41286%201.66667%201.66667%202.41286%201.66667%203.33333V13.3333C1.66667%2014.2538%202.41286%2015%203.33333%2015H5.83333V18.3333L10.8333%2015H16.6667C17.5871%2015%2018.3333%2014.2538%2018.3333%2013.3333V3.33333C18.3333%202.41286%2017.5871%201.66667%2016.6667%201.66667Z'%20fill='white'%20stroke='%2364A4F4'%20strokeWidth='2'%20strokeLinecap='round'%20strokeLinejoin='round'/%3e%3c/svg%3e"
      />
    </div>
    <div
      class="flex-1"
    >
      <div
        class="flex items-center gap-2"
      >
        <h3
          class="lg:text-sm text-[12px] font-medium leading-5 text-gray-600"
        >
          Test notification
        </h3>
      </div>
      <p
        class="text-xs sm:text-sm text-gray-500 mt-0.5"
      >
        2 minutes ago
      </p>
    </div>
    <div
      class="flex items-center gap-1"
    >
      <button
        class="p-1 sm:p-2 text-gray-400 hover:text-blue-600 rounded-full transition-colors"
        title="Mark as unread"
      >
        <svg
          class="w-3 h-3 sm:w-4 sm:h-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M2.458 12C3.732 7.943 7.523 5 12 5c4.478 0 8.268 2.943 9.542 7-1.274 4.057-5.064 7-9.542 7-4.477 0-8.268-2.943-9.542-7z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
      <button
        class="p-1 sm:p-2 text-gray-400 hover:text-red-500 rounded-full transition-colors"
        title="Delete notification"
      >
        <svg
          class="w-3 h-3 sm:w-4 sm:h-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`NotificationCard Component > matches snapshot for unread notification 1`] = `
<div
  class="py-2 sm:py-3 first:pt-0 transition-all duration-200 opacity-100"
>
  <div
    class="flex items-start"
  >
    <div
      class="lg:w-10 lg:h-10 md:w-8 md:h-8 w-6 h-6 rounded-lg flex items-center justify-center mr-3"
      style="background-color: rgb(100, 164, 244);"
    >
      <img
        alt="info notification"
        class="lg:w-5 lg:h-5 md:w-4 md:h-4 w-3 h-3"
        src="data:image/svg+xml,%3c?xml%20version='1.0'%20encoding='UTF-8'?%3e%3csvg%20width='20'%20height='20'%20viewBox='0%200%2020%2020'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M16.6667%201.66667H3.33333C2.41286%201.66667%201.66667%202.41286%201.66667%203.33333V13.3333C1.66667%2014.2538%202.41286%2015%203.33333%2015H5.83333V18.3333L10.8333%2015H16.6667C17.5871%2015%2018.3333%2014.2538%2018.3333%2013.3333V3.33333C18.3333%202.41286%2017.5871%201.66667%2016.6667%201.66667Z'%20fill='white'%20stroke='%2364A4F4'%20strokeWidth='2'%20strokeLinecap='round'%20strokeLinejoin='round'/%3e%3c/svg%3e"
      />
    </div>
    <div
      class="flex-1"
    >
      <div
        class="flex items-center gap-2"
      >
        <h3
          class="lg:text-sm text-[12px] font-medium leading-5 text-gray-900"
        >
          Test notification
        </h3>
        <div
          class="w-2 h-2 bg-blue-500 rounded-full flex-shrink-0"
          data-testid="unread-indicator"
        />
      </div>
      <p
        class="text-xs sm:text-sm text-gray-500 mt-0.5"
      >
        2 minutes ago
      </p>
    </div>
    <div
      class="flex items-center gap-1"
    >
      <button
        class="p-1 sm:p-2 text-gray-400 hover:text-blue-600 rounded-full transition-colors"
        title="Mark as read"
      >
        <svg
          class="w-3 h-3 sm:w-4 sm:h-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.875 18.825A10.05 10.05 0 0112 19c-4.478 0-8.268-2.943-9.543-7a9.97 9.97 0 011.563-3.029m5.858.908a3 3 0 114.243 4.243M9.878 9.878l4.242 4.242M9.878 9.878L3 3m6.878 6.878L21 21"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
      <button
        class="p-1 sm:p-2 text-gray-400 hover:text-red-500 rounded-full transition-colors"
        title="Delete notification"
      >
        <svg
          class="w-3 h-3 sm:w-4 sm:h-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
